src/Adapter/Discount/Update/DiscountDuplicator.php

```diff
@@ -0,0 +1,331 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Discount\Update;
+
+use CartRule;
+use Doctrine\DBAL\ArrayParameterType;
+use Doctrine\DBAL\Connection;
+use Language;
+use PrestaShop\PrestaShop\Adapter\Discount\Repository\DiscountRepository;
+use PrestaShop\PrestaShop\Core\Domain\Discount\DiscountSettings;
+use PrestaShop\PrestaShop\Core\Domain\Discount\Exception\CannotDuplicateDiscountException;
+use PrestaShop\PrestaShop\Core\Domain\Discount\ValueObject\DiscountId;
+use PrestaShop\PrestaShop\Core\Hook\HookDispatcherInterface;
+use PrestaShop\PrestaShop\Core\Repository\AbstractObjectModelRepository;
+use PrestaShop\PrestaShop\Core\Util\String\StringModifierInterface;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+/**
+ * Duplicates discount
+ */
+class DiscountDuplicator extends AbstractObjectModelRepository
+{
+    public function __construct(
+        protected readonly DiscountRepository $discountRepository,
+        protected readonly HookDispatcherInterface $hookDispatcher,
+        protected readonly Connection $connection,
+        protected readonly string $dbPrefix,
+        protected readonly TranslatorInterface $translator,
+        protected readonly StringModifierInterface $stringModifier,
+    ) {
+    }
+
+    /**
+     * Global process of duplicating a discount
+     */
+    public function duplicate(DiscountId $discountId): DiscountId
+    {
+        $oldDiscountId = $discountId->getValue();
+        $this->hookDispatcher->dispatchWithParameters(
+            'actionAdminDuplicateBefore',
```

Comment:
```suggestion
            'actionAdminDuplicateDiscountBefore',
```

I know this was copied from `ProductDuplicator`, but now I think the hook name for product was not accurate enough when I see this new example.

We could tell the two cases apart based on the parameters (id_discount instead of id_product), but it may not be ideal 🤔